File: packages/core/canvas/duc/duc-ts/src/serializeDuc.ts
import * as flatbuffers from 'flatbuffers';
import { AppState as BinAppState, BinaryFiles as BinBinaryFiles, ExportedDataState, BinaryFilesEntry, BinaryFileData, DucElement as BinDucElement } from '../duc';
import { fileSave } from '../../../data/filesystem';
import { DEFAULT_FILENAME, EXPORT_DATA_TYPES, EXPORT_SOURCE, MIME_TYPES, VERSIONS } from '../../../constants';
import { DucElement } from '../../../element/types';
import { AppState, BinaryFiles } from '../../../types';
import { serializeDucElement } from './serialize/serializeElementFromDuc';
import { serializeAppState } from './serialize/serializeAppStateFromDuc';
import { serializeBinaryFiles } from './serialize/serializeBinaryFilesFromDuc';
import { restore } from '../../../data/restore';

export const serializeAsFlatBuffers = async (
  elements: readonly DucElement[],
  appState: Partial<AppState>,
  files: BinaryFiles,
): Promise<Uint8Array> => {
  const builder = new flatbuffers.Builder(1024);

  const sanitized = restore(
    {
      elements,
      appState,
      files,
    }, null, null, {
      refreshDimensions: false,
    }
  );

  // Serialize elements
  const elementOffsets = sanitized.elements.map((element) => {
    return serializeDucElement(builder, element);
  });
  const elementsOffset = ExportedDataState.createElementsVector(builder, elementOffsets);

  // Serialize appState
  const appStateOffset = serializeAppState(builder, sanitized.appState);

  // Serialize files
  const binaryFilesOffset = serializeBinaryFiles(builder, sanitized.files);

  // Serialize ExportedDataState
  const typeOffset = builder.createString(EXPORT_DATA_TYPES.duc);
  const sourceOffset = builder.createString(EXPORT_SOURCE);

  ExportedDataState.startExportedDataState(builder);
  ExportedDataState.addType(builder, typeOffset);
  ExportedDataState.addVersion(builder, VERSIONS.duc);
  ExportedDataState.addSource(builder, sourceOffset);
  ExportedDataState.addElements(builder, elementsOffset);
  ExportedDataState.addAppState(builder, appStateOffset);
  ExportedDataState.addFiles(builder, binaryFilesOffset);
  const exportedDataStateOffset = ExportedDataState.endExportedDataState(builder);

  builder.finish(exportedDataStateOffset);

  return builder.asUint8Array();
};


export const saveAsFlatBuffers = async (
  elements: readonly DucElement[],
  appState: AppState,
  files: BinaryFiles,
  name: string = DEFAULT_FILENAME,
) => {
  try {
    const serialized = await serializeAsFlatBuffers(elements, appState, files);
    const blob = new Blob([serialized], {
      type: MIME_TYPES.duc,
    });
  
    const fileHandle = await fileSave(blob, {
      name,
      extension: "duc",
      description: "Duc file",
    });
    return { fileHandle };
  } catch (error) {
    console.error(error);
    throw new Error("Failed to save file");
  }
};

File: packages/core/canvas/duc/duc-py/src/serialize_duc.py
import flatbuffers
from Duc.ExportedDataState import *
from .serialize.serialize_duc_element import serialize_duc_element
from .serialize.serialize_app_state import serialize_app_state
from .serialize.serialize_binary_files import serialize_binary_files
from .constants import EXPORT_DATA_TYPES, EXPORT_SOURCE, VERSIONS
from .models.AppState import AppState
from .models.DucElement import DucElementUnion
from .models.BinaryFiles import BinaryFiles

def serialize_as_flatbuffers(elements: list[DucElementUnion], app_state: AppState, files: BinaryFiles, export_type: str) -> bytes:
    builder = flatbuffers.Builder(1024)

    # Serialize elements
    element_offsets = [serialize_duc_element(builder, element) for element in elements]
    elements_vector = ExportedDataStateStartElementsVector(builder, element_offsets)

    # Serialize appState
    app_state_offset = serialize_app_state(builder, app_state)

    # Serialize files
    binary_files_offset = serialize_binary_files(builder, files)

    # Serialize ExportedDataState
    type_offset = builder.CreateString(EXPORT_DATA_TYPES['duc'])
    source_offset = builder.CreateString(EXPORT_SOURCE)

    ExportedDataStateStart(builder)
    ExportedDataStateAddType(builder, type_offset)
    ExportedDataStateAddVersion(builder, VERSIONS['excalidraw'])
    ExportedDataStateAddSource(builder, source_offset)
    ExportedDataStateAddElements(builder, elements_vector)
    ExportedDataStateAddAppState(builder, app_state_offset)
    ExportedDataStateAddFiles(builder, binary_files_offset)
    exported_data_state_offset = ExportedDataStateEnd(builder)

    builder.Finish(exported_data_state_offset)

    return builder.Output()

def save_as_flatbuffers(elements: list[DucElementUnion], app_state: AppState, files: BinaryFiles, name: str = "Untitled") -> bytes:
    serialized = serialize_as_flatbuffers(elements, app_state, files, "local")
    return serialized
